File: 2/solution.js
import { readFileSync } from "fs";

const input_list = readFileSync("2/data.txt", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.split(/\s+/).map(Number));

// * Part 1

const is_increasing = (report) => {
  for (let i = 0; i < report.length - 1; i++) {
    if (report[i] < report[i + 1]) continue;
    else return false;
  }
  return true;
};

const is_decreasing = (report) => {
  for (let i = 0; i < report.length - 1; i++) {
    if (report[i] > report[i + 1]) continue;
    else return false;
  }
  return true;
};

const is_between_1_and_3 = (report) => {
  for (let i = 0; i < report.length - 1; i++) {
    if (
      Math.abs(report[i + 1] - report[i]) >= 1 &&
      Math.abs(report[i + 1] - report[i]) <= 3
    )
      continue;
    else return false;
  }
  return true;
};

const is_safe = (report) => {
  if (
    (is_increasing(report) || is_decreasing(report)) &&
    is_between_1_and_3(report)
  )
    return true;
  else return false;
};

let safe = 0;

for (let i = 0; i < input_list.length; i++) {
  if (is_safe(input_list[i])) safe++;
}

console.log(safe);

// * Part 2

const is_monotonic_and_in_range = (levels) => {
  let diffs = [];
  for (let i = 0; i < levels.length - 1; i++) {
    diffs.push(levels[i] - levels[i + 1]);
  }

  let is_monotonic = true;
  for (let i = 0; i < diffs.length; i++) {
    if (diffs[i] <= 0) {
      is_monotonic = false;
      break;
    }
  }
  if (!is_monotonic) {
    is_monotonic = true;
    for (let i = 0; i < diffs.length; i++) {
      if (diffs[i] >= 0) {
        is_monotonic = false;
        break;
      }
    }
  }

  let is_in_range = true;
  for (let i = 0; i < diffs.length; i++) {
    if (Math.abs(diffs[i]) < 1 || Math.abs(diffs[i]) > 3) {
      is_in_range = false;
      break;
    }
  }

  if (is_monotonic && is_in_range) return true;
  return false;
};

safe = 0;

for (let report of input_list) {
  if (is_monotonic_and_in_range(report)) {
    safe++;
  } else {
    for (let i = 0; i < report.length; i++) {
      let tolerated_levels = report.slice(0, i).concat(report.slice(i + 1));
      if (is_monotonic_and_in_range(tolerated_levels)) {
        safe++;
        break;
      }
    }
  }
}

console.log(safe);

File: 2/solution.java
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class solution {
  public static void main(String[] args) throws IOException {
    List<List<Integer>> inputList = new ArrayList<>();
    try (BufferedReader br = new BufferedReader(new FileReader("data.txt"))) {
      String line;
      while ((line = br.readLine()) != null) {
        List<Integer> report = Arrays.stream(line.trim().split("\\s+"))
                       .map(Integer::parseInt)
                       .collect(Collectors.toList());
        inputList.add(report);
      }
    }

    // Part 1
    int safe = 0;
    for (List<Integer> report : inputList) {
      if (isSafe(report)) {
        safe++;
      }
    }
    System.out.println(safe);

    // Part 2
    safe = 0;
    for (List<Integer> report : inputList) {
      if (isMonotonicAndInRange(report)) {
        safe++;
      } else {
        for (int i = 0; i < report.size(); i++) {
          List<Integer> toleratedLevels = new ArrayList<>(report);
          toleratedLevels.remove(i);
          if (isMonotonicAndInRange(toleratedLevels)) {
            safe++;
            break;
          }
        }
      }
    }
    System.out.println(safe);
  }

  private static boolean isIncreasing(List<Integer> report) {
    for (int i = 0; i < report.size() - 1; i++) {
      if (report.get(i) >= report.get(i + 1)) {
        return false;
      }
    }
    return true;
  }

  private static boolean isDecreasing(List<Integer> report) {
    for (int i = 0; i < report.size() - 1; i++) {
      if (report.get(i) <= report.get(i + 1)) {
        return false;
      }
    }
    return true;
  }

  private static boolean isBetween1And3(List<Integer> report) {
    for (int i = 0; i < report.size() - 1; i++) {
      int diff = Math.abs(report.get(i + 1) - report.get(i));
      if (diff < 1 || diff > 3) {
        return false;
      }
    }
    return true;
  }

  private static boolean isSafe(List<Integer> report) {
    return (isIncreasing(report) || isDecreasing(report)) && isBetween1And3(report);
  }

  private static boolean isMonotonicAndInRange(List<Integer> levels) {
    List<Integer> diffs = new ArrayList<>();
    for (int i = 0; i < levels.size() - 1; i++) {
      diffs.add(levels.get(i) - levels.get(i + 1));
    }

    boolean isMonotonic = true;
    for (int diff : diffs) {
      if (diff <= 0) {
        isMonotonic = false;
        break;
      }
    }
    if (!isMonotonic) {
      isMonotonic = true;
      for (int diff : diffs) {
        if (diff >= 0) {
          isMonotonic = false;
          break;
        }
      }
    }

    boolean isInRange = true;
    for (int diff : diffs) {
      if (Math.abs(diff) < 1 || Math.abs(diff) > 3) {
        isInRange = false;
        break;
      }
    }

    return isMonotonic && isInRange;
  }
}